supportconfig: examine your infrastructure in just one go

isbm opened the pull request:
### What does this PR do?

- Allows to run `salt-support` command on the Minion remotely and via SaltSSH[1].
- Adds state runner

NOTE: Documentation to Salt Support and module/state usage is coming in a separate doc-only PR.

### General Usage

If you are familiar with the Salt Support feature, you already know that you can just call `salt-support` command in CLI and it will examine your _current_ Salt installed components. Now this is possible to do on remote minions.

- get list of existing archives: `support.archives`
- delete archives: `support.delete_archives` 
- get the last available archive `support.last_archive`
- list available profiles on the minion: `support.profiles`
- Run the support: `support.run`
- Sync the latest archive to the host on given location: `support.sync`

It goes as usual, e.g.: `salt \* support.run profile=network`

### State Usage

Now, if you can collect the support data from the minions that matches your criteria, how about collecting all that data in one central place? In that way you can just examine your entire cluster in just one go:

```yaml
examine_my_systems:
  support:
    - taken:
      - profile: network
      - output: nested

    - collected:
      - group: my-cluster
```

The state above will call support on all the machines you're applying this state collecting Networking data using `network` profile and writing it down using `nested` outputter. Then it will collect it under the `/tmp/my-cluster/` directory.

### Pillars

You can configure your own profile scenario in pillar under `support_config` section and point it out right away from the master. For example:

```yaml
support_config:
  pillar_logfile:  # <-- Point this profile ID!
    - filetree:
        info: Add current logfile, by pillar
        args:
          - /var/log/messages
          - /var/log/syslog
          - {{ salt['config.get']('log_file') }}
```

### Tests written?

Of course. :wink: 

_________________
1. This is still limited to the SaltSSH-copied instance, i.e. no examination of system-installed Salt components.

Commit:
Add initial scaffold for the saltsupport module

Commit:
Add log collector for remote purposes

Commit:
Implement default archive name

Commit:
Fix imports

Commit:
Implement runner function

Commit:
Remove targets data collector function as it is now called by a module instead

Commit:
Add external method decorator marker

Commit:
Add utility class for detecting exportable methods

Commit:
Mark run method as an external function

Commit:
Implement function setter

Commit:
Fix imports

Commit:
Setup config from __opts__

Commit:
Use utility class

Commit:
Remove utils class

Commit:
Allow specify profile from the API parameter directly

Commit:
Rename module by virtual name

Commit:
Bypass parent subclass

Commit:
Implement profiles listing (local only for now)

Commit:
Specify profile from the state/call

Commit:
Set default or personalised archive name

Commit:
Add archives lister

Commit:
Add personalised name element to the archive name

Commit:
Use proper args/kwargs to the exported function

Commit:
Add archives deletion function

Commit:
Change log level when debugging rendered profiles

Commit:
Add ability to directly pass profile source when taking local data

Commit:
Add pillar profile support

Commit:
Remove extra-line

Commit:
Fix header

Commit:
Change output format for deleting archives

Commit:
Refactor logger output format

Commit:
Add time/milliseconds to each log notification

Commit:
Fix imports

Commit:
Switch output destination by context

Commit:
Add last archive function

Commit:
Lintfix

Commit:
Return consistent type

Commit:
Change output format for deleted archives report

Commit:
Implement report archive syncing to the reporting node

Commit:
Send multiple files at once via rsync, instead of send one after another

Commit:
Add sync stats formatter

Commit:
Change signature: cleanup -> move. Update docstring.

Commit:
Flush empty data from the output format

Commit:
Report archfiles activity

Commit:
Refactor imports

Commit:
Do not remove retcode if it is EX_OK

Commit:
Do not raise rsync error for undefined archives.

Commit:
Update header

Commit:
Add salt-support state module

Commit:
Move all functions into a callable class object

Commit:
Support __call__ function in state and command modules as default entrance that does not need to be specified in SLS state syntax

Commit:
Access from the outside only allowed class methods

Commit:
Pre-create destination of the archive, preventing single archive copied as a group name

Commit:
Handle functions exceptions

Commit:
Add unit test scaffold

Commit:
Add LogCollector UT for testing regular message

Commit:
Add LogCollector UT for testing INFO message

Commit:
Add LogCollector UT for testing WARNING message

Commit:
Replace hardcoded variables with defined constants

Commit:
Add LogCollector UT for testing ERROR message

Commit:
Test title attribute in msg method of LogCollector

Commit:
Add UT for LogCollector on highlighter method

Commit:
Add UT for LogCollector on put method

Commit:
Fix docstrings

Commit:
Add UT for archive name generator

Commit:
Add UT for custom archive name

Commit:
Fix docstring for the UT

Commit:
Add UT for checking profiles list format

Commit:
Add Unit Test for existing archives listing

Commit:
Add UT for the last archive function

Commit:
Create instance of the support class

Commit:
Add UT for successfully deleting all archives

Commit:
Add UT for deleting archives with failures

Commit:
Add UI for formatting sync stats and order preservation

Commit:
Add UT for testing sync failure when no archives has been specified

Commit:
Add UT for last picked archive has not found

Commit:
Add UT for last specified archive was not found

Commit:
Bugfix: do not create an array with None element in it

Commit:
Fix UT for found bugfix

Commit:
Add UT for syncing no archives failure

Commit:
Add UT for sync function

Commit:
Add UT for run support function

isbm commented:
@cachedout here you go. :wink:

Commit:
Fix docstring for function "run"

Commit:
lintfix: use 'salt.support.mock' and 'patch()'

Commit:
Rewrite subdirectory creation and do not rely on Python3-only code

Commit:
Lintfix: remove unused imports

Commit:
Lintfix: regexp strings

cachedout commented:
@isbm Nice!

Can you grab the lint errors to start? https://jenkinsci.saltstack.com/job/pr-lint/job/PR-50401/3/warnings52Result/new/

isbm commented on salt/loader.py:
```diff
@@ -1679,8 +1679,10 @@ def _load_module(self, name):
+            if attr.startswith('_') and attr != '__call__':
+                # private functions are skipped,
+                # except __call__ which is default entrance
+                # for multi-function batch-like state syntax
```
This allows modules to have optional `__call__` function which is hidden in the module. So instead of:

```yaml
my_something_to_do:
  somemodule.function:
    - param

my_something_to_do_part_two:
  somemodule.otherfunction:
     - param

my_something_why_do_i_keep_doing_this:
  somemodule.thirdfunction:
    - param

my_something_oh_stop_that_please:
  somemodule.last_function:
    - param
```

You can now do it much simpler:
```yaml
my_something_to_do:
  somemodule:
    - function:
      - param
    - otherfunction:
      - param
    - last_function:
      - param

```
If you do not define function after `somemodule`, then in this case `__call__` will be injected to `somemodule.__call__` and it will iterate over the list (order preserved) introspecting module and calling things one by one.

Both syntaxes in parallel can perfectly exist.

isbm replied:
Every function with underscore is yanked from the module to get exported. Which means "private". I personally don't think this is the best solution, because it is impossible to keep hidden function that does not starts from underscore. In Zope, GvR made another convention: no `__doc__` = stays private. Anyway, only `__call__` is allowed. It (conceptually) works this way:

```python
def __call__(*function_names, **params):
    ret = []
    for func in function_names:
        ret.append(getattr(this_module, func)(*params[func]['args'], **params[func]['kwargs']))
    return ret
```

isbm commented on salt/modules/saltsupport.py:
```diff
@@ -0,0 +1,372 @@
+    for m_name in dir(support):
+        obj = getattr(support, m_name)
+        if getattr(obj, 'external', False):
+            setattr(sys.modules[__name__], m_name, _set_function(obj))
```
The module is implemented as a class. So _some_ class's methods are exported as module functions, since decorator of those sets these methods as external.

isbm commented on salt/state.py:
```diff
@@ -2783,10 +2783,31 @@ def call_listen(self, chunks, running):
         '''
         Process a high data call and ensure the defined states.
         '''
+        self.inject_default_call(high)
```
Do you have evidence we aren't careful? :wink: It actually doesn't clash at all. Performance wise it hardly adds even a nanofraction of a second per minion. All what it does, just looks if there is function defined on top level (and usually it is). If you did not defined one, it adds `__call__` reference and quits. Or just quits (default). BTW, do we have already some numbers that proves it affects performance?

But on the other hand this it allows to get rid of ID-per-function SLS style that forces to write repetitive states, yet perfectly preserving original way without clashes.

isbm replied:
@cachedout some update: I've measured performance. It is actually the same, except new version (single-id) in some very small fraction a bit faster in average. But I don't know how true this is, since many factors are involved.

Commit:
Break-down oneliner if/else clause

Commit:
Use ordered dictionary to preserve order of the state.

This has transparent effect to the current process: OrderedDict is the
same as just Python dict, except it is preserving order of the state
chunks.

Commit:
Refactor state processing class.

Commit:
Add __call__ function to process single-id syntax

Commit:
Add backward-compatibility with default SLS syntax (id-per-call)

Commit:
Lintfix: E1120 no value in argument 'name' for class constructor

Commit:
Remove unused import

Commit:
Check last function by full name

isbm commented:
@cachedout this is [the RFC you were asking for](https://github.com/isbm/salt/blob/isbm-rfc-simpler-sls-syntax/rfcs/sls-alternative-syntax.md).

thatch45 commented on salt/modules/saltsupport.py:
```diff
@@ -0,0 +1,381 @@
+        self.msg(message.format(*values))
+
+
+class SaltSupportModule(SaltSupport):
```
Why not just write functions and use `__context__`? Like the rest of salt, why do you need classes here? python != Java

thatch45 reviewed (changes requested):
I think this is way too OOP, we can do this work with far less code if it was rewritten in a functional and more salt like manner

thatch45 commented:
If we can get this branch updated and the tests looking clean then we can move forward here.

isbm commented:
@thatch45 updated the branch, let tests run and I will get back to this soon.

isbm commented:
Codeclimate is failing, but AFAIK we had that discussion with @cachedout before and this seems wasn't yet fixed. So I'd skip this failure, apparently.

terminalmage commented:
ported to master in https://github.com/saltstack/salt/pull/57054
